File: data-structures/Algorithms/mergeSort.py
def merge(nums, low, mid, high):
    temp = []
    left = low
    right = mid+1
    while(left <=mid and right<=high):
        if nums[left] <= nums[right]:
            temp.append(nums[left])
            left += 1
        else:
            temp.append(nums[right])
            right += 1
    
    while(left <= mid):
        temp.append(nums[left])
        left += 1
    
    while(right <= high):
        temp.append(nums[right])
        right += 1
    
    # inserting from temp to original array
    for i in range(low, high+1):
        nums[i] = temp[i-low]

def mergeSort(nums, low, high):
    if low >= high:
        return
    
    mid = (low+high)//2
    mergeSort(nums, low, mid)
    mergeSort(nums, mid+1, high)
    merge(nums, low, mid, high)



nums = [5,2,6,4,3,1]
mergeSort(nums, 0, len(nums)-1)

nums2 = [324,1342,354,34,34,3243,545,0]
mergeSort(nums2, 0, len(nums2)-1)
print(nums2)

File: algorithms-and-sums/Sorting/mergeSort.cpp
#include<bits/stdc++.h>

using namespace std;

void merge(vector<int> &arr, int left, int mid, int right){
	int low = left;
	int high = mid+1;
	vector<int> temp;
	while(low <= mid && high <=right){
		if(arr[low] <= arr[high]){
			temp.push_back(arr[low]);
			low++;
		}else{
			temp.push_back(arr[high]);
			high++;
		}
	}
	
	while(low <= mid){
		temp.push_back(arr[low]);
		low++;
	}

	while(high <= right){
		temp.push_back(arr[high]);
		high++;
	}

	for(int i=left; i<=right; i++){
		arr[i] = temp[i-left];
	}
}

void mergeSort(vector<int> &arr, int left, int right){
	if(left >= right){
		return;
	}
	int mid = (left + right)/2;
	mergeSort(arr, left, mid);
	mergeSort(arr, mid+1, right);
	merge(arr, left, mid, right);
}

template <typename T> void print(vector<T> arr){
    for(auto num: arr){
        cout << num << " ";
    }
    cout << endl;
}

int main(){
    vector<int> arr = {1,0,5,2,4,5,26,2,63,66,3,0,45,2,3};
    mergeSort(arr, 0, arr.size()-1);
    print(arr);
}